src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -8,30 +8,128 @@ namespace Microsoft.Azure.ServiceBus.Core
     using System.Linq;
     using System.Threading;
     using System.Threading.Tasks;
+    using Microsoft.Azure.Amqp;
+    using Microsoft.Azure.Amqp.Encoding;
+    using Microsoft.Azure.Amqp.Framing;
+    using Microsoft.Azure.ServiceBus.Amqp;
+    using Microsoft.Azure.ServiceBus.Primitives;
 
-    public abstract class MessageReceiver : ClientEntity, IMessageReceiver
+    public class MessageReceiver : ClientEntity, IMessageReceiver
     {
-        readonly TimeSpan operationTimeout;
+        public static readonly TimeSpan DefaultBatchFlushInterval = TimeSpan.FromMilliseconds(20);
+        private const int DefaultPrefetchCount = 0;
+
+        readonly ConcurrentExpiringSet<Guid> requestResponseLockedMessages;
+        readonly bool isSessionReceiver;
         readonly object messageReceivePumpSyncLock;
+        readonly bool ownsConnection;
+
         int prefetchCount;
         long lastPeekedSequenceNumber;
         MessageReceivePump receivePump;
         CancellationTokenSource receivePumpCancellationTokenSource;
 
+        public MessageReceiver(
+            ServiceBusConnectionStringBuilder connectionStringBuilder,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : this(connectionStringBuilder.GetNamespaceConnectionString(), connectionStringBuilder.EntityPath, receiveMode, retryPolicy)
+        {
+        }
+
+        public MessageReceiver(
+            string connectionString,
+            string entityPath,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : this(entityPath, new ServiceBusNamespaceConnection(connectionString), DefaultPrefetchCount, receiveMode, retryPolicy)
+        {
+            if (string.IsNullOrWhiteSpace(entityPath))
+            {
+                throw Fx.Exception.ArgumentNullOrWhiteSpace(entityPath);
+            }
+
+            this.ownsConnection = true;
+        }
+
+        public MessageReceiver(
+            string entityPath,
+            ServiceBusConnection serviceBusConnection,
+            int prefetchCount = DefaultPrefetchCount,
+            ReceiveMode receiveMode = ReceiveMode.PeekLock,
+            RetryPolicy retryPolicy = null)
+            : this(entityPath, null, receiveMode, serviceBusConnection, null, retryPolicy)
+        {
+            var tokenProvider = TokenProvider.CreateSharedAccessSignatureTokenProvider(serviceBusConnection.SasKeyName, serviceBusConnection.SasKey);
+            this.CbsTokenProvider = new TokenProviderAdapter(tokenProvider, serviceBusConnection.OperationTimeout);
+        }
+
+        internal MessageReceiver(
```

Comment:
Left a single internal constructor, and unit tests all use public constructors now